File: Algorithm/C++/1000+/1020. Partition Array Into Three Parts With Equal Sum.cc
/*
 * Created on Thu Mar 28 2019 23:33:46
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

#include <vector>
#include <numeric>

class Solution {
 public:
  bool canThreePartsEqualSum(std::vector<int>& A) {
    int sum = std::accumulate(A.begin(), A.end(), 0);
    if (sum % 3 != 0) return false;
    int ave = sum/3;
    int count = 0, tmp = 0;
    for (auto num : A) {
      if (tmp == ave) {
        count++;
        tmp = 0;
      }
      tmp += num;
    }
    if (tmp == ave) count++;
    return count==3;
  }
};

File: Algorithm/Go/1000+/1020. Partition Array Into Three Parts With Equal Sum.go
/*
 * Created on Thu Mar 28 2019 23:39:19
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

func canThreePartsEqualSum(A []int) bool {
	s := 0
	for _, num := range A {
		s += num
	} 
	if s % 3 != 0 {
		return false
	}
	ave := s/3
	count, tmp := 0, 0
	for _, num := range A {
		if tmp == ave {
			count++
			tmp = 0
		}
		tmp += num
	}
	if tmp == ave {
		count++
	}
	return count == 3
}
